spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-yarn/src/main/java/org/springframework/cloud/data/module/deployer/yarn/YarnModuleDeployer.java

```diff
@@ -0,0 +1,213 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.data.module.deployer.yarn;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.data.core.ModuleCoordinates;
+import org.springframework.cloud.data.core.ModuleDefinition;
+import org.springframework.cloud.data.core.ModuleDeploymentId;
+import org.springframework.cloud.data.core.ModuleDeploymentRequest;
+import org.springframework.cloud.data.module.ModuleStatus;
+import org.springframework.cloud.data.module.deployer.ModuleDeployer;
+import org.springframework.util.StringUtils;
+import org.springframework.yarn.boot.app.YarnContainerClusterApplication;
+import org.springframework.yarn.boot.app.YarnInfoApplication;
+
+/**
+ * {@link ModuleDeployer} which communicates to XD's yarn app running
+ * on a hadoop cluster waiting for deployment requests. This app is
+ * utilising spring yarn's container grouping functionality to create a
+ * new group per module type which then allows all modules to share same
+ * settings and group itself can controlled, i.e. ramp up/down or shutdown/destroy
+ * whole group.
+ *
+ * @author Janne Valkealahti
+ */
+public class YarnModuleDeployer implements ModuleDeployer {
+
+	private static final Logger logger = LoggerFactory.getLogger(YarnModuleDeployer.class);
+	private static final String PREFIX = "spring.yarn.internal.ContainerClusterApplication.";
+
+	public YarnModuleDeployer() {
+	}
+
+	@Override
+	public ModuleDeploymentId deploy(ModuleDeploymentRequest request) {
+		int count = request.getCount();
+		ModuleCoordinates coordinates = request.getCoordinates();
+		ModuleDefinition definition = request.getDefinition();
+		logger.info("deploying request for definition: " + definition);
+
+		ModuleDeploymentId moduleDeploymentId = new ModuleDeploymentId(definition.getGroup(), definition.getLabel());
+		String clusterId = sanitizeClusterId(convertFromModuleDeploymentId(moduleDeploymentId));
+
+		String yarnApplicationId = findRunningXdYarnApp();
+		logger.info("Using application id " + yarnApplicationId);
+		String module = coordinates.toString();
+		logger.info("deploying module: " + module);
+
+		Map<String, String> definitionParameters = definition.getParameters();
+		Map<String, String> deploymentProperties = request.getDeploymentProperties();
+		logger.info("definitionParameters: " + definitionParameters);
+		logger.info("deploymentProperties: " + deploymentProperties);
+
+		// Using same app instance yarn boot cli is using to
+		// communicate with an app running on yarn via its boot actuator
+		YarnContainerClusterApplication app = new YarnContainerClusterApplication();
+		Properties appProperties = new Properties();
+		appProperties.setProperty(PREFIX + "operation", "CLUSTERCREATE");
```

Comment:
Yes, cluster/group is for each module which is what diego(in a sense would do) so that we can ramp up/down modules, or stop/start/destroy all similar modules.